homeassistant/components/enphase_envoy/number.py

```diff
@@ -0,0 +1,126 @@
+"""Number platform for Enphase Envoy solar energy monitor."""
+from __future__ import annotations
+
+from collections.abc import Callable
+from dataclasses import dataclass
+import logging
+
+from pyenphase import EnvoyDryContactSettings
+
+from homeassistant.components.number import (
+    NumberDeviceClass,
+    NumberEntity,
+    NumberEntityDescription,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import EntityCategory
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.device_registry import DeviceInfo
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .const import DOMAIN
+from .coordinator import EnphaseUpdateCoordinator
+from .entity import EnvoyBaseEntity
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@dataclass
+class EnvoyRelayRequiredKeysMixin:
+    """Mixin for required keys."""
+
+    value_fn: Callable[[EnvoyDryContactSettings], float]
+
+
+@dataclass
+class EnvoyRelayNumberEntityDescription(
+    NumberEntityDescription, EnvoyRelayRequiredKeysMixin
+):
+    """Describes an Envoy Dry Contact Relay number entity."""
+
+
+RELAY_ENTITIES = (
+    EnvoyRelayNumberEntityDescription(
+        key="soc_low",
+        translation_key="cutoff_battery_level",
+        device_class=NumberDeviceClass.BATTERY,
+        entity_category=EntityCategory.CONFIG,
+        value_fn=lambda relay: relay.soc_low,
+    ),
+    EnvoyRelayNumberEntityDescription(
+        key="soc_high",
+        translation_key="restore_battery_level",
+        device_class=NumberDeviceClass.BATTERY,
+        entity_category=EntityCategory.CONFIG,
+        value_fn=lambda relay: relay.soc_high,
+    ),
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up Enphase Envoy number platform."""
+    coordinator: EnphaseUpdateCoordinator = hass.data[DOMAIN][config_entry.entry_id]
+    envoy_data = coordinator.envoy.data
+    assert envoy_data is not None
+    envoy_serial_num = config_entry.unique_id
+    assert envoy_serial_num is not None
```

Comment:
in which case is `envoy_serial_num` None ?

Reply:
Actually don't we set it in the base entity class now?

Reply:
Cleaned up the rest here https://github.com/home-assistant/core/pull/98474